omgnetwork/elixir-omg
# feat: block queue metrics and stalled submission alarm

Closes #1629

## Overview

Publish various block queue metrics (see the list in Changes section) and `block_submit_stalled` alarm.

## Changes

### Telemetry
- Publishes `:blknum_submitting` telemetry event with the `blknum` when a block is about to be submitted
- Publishes `:blknum_submitted` telemetry event with the `blknum` when a block is successfully submitted

These are used for metrics reporting.

### OMG.Bus
- Broadcasts `:block_submitting` event with the `blknum` when a block is about to be submitted
- Broadcasts `:block_submitted` event with the `blknum` when a block is successfully submitted

These are used by `BlockQueue.Monitor` to keep tracking of stalled blocks.

### Datadog metrics
- `block_queue_blknum_submitting`: Reports the blknum being attempted (e.g. 1000, 1000, 2000, 1000, 2000, 3000, 2000, 3000, ...)
- `block_queue_blknum_submitted`: Reports the blknum succesfully submitted (e.g. 1000, 2000, 3000, ...)
- `block_submission_attempt`: Increments by 1 when a block submission is attempted
- `block_submission_success`: Increments by 1 when a block submission is successfully submitted
- `block_queue_num_blocks_stalled`: Reports the number of stalled blocks at the moment of metric collection (e.g. 0, 0, 2, 1, 0, 1, 0, ...)
- Rename `block_submission` gauge to `block_submission_gas` for better naming accuracy

### Alarm
- Raises `:block_submit_stalled` alarm when at least one block has been submitting for the past `BLOCK_SUBMIT_STALL_THRESHOLD_BLOCKS` root chain blocks or longer
- Clears `:block_submit_stalled` alarm when all submitting blocks have been submitting for less than `BLOCK_SUBMIT_STALL_THRESHOLD_BLOCKS` root chain blocks
- Defaults `BLOCK_SUBMIT_STALL_THRESHOLD_BLOCKS` env var to 4 root chain blocks

The threshold is in number of root chain blocks because blocktime can vary greatly between networks. Even on mainnet blocktime can vary between ~ 1-60 seconds, and a block submission should not be considered stalled, for example, after 30 seconds of submitting if the blocktime is 45 seconds.

### Miscellaneous
- Change log level for block formation from debug to info

## Testing

1. Set geth's `--miner.gasprice` to a very high value, e.g. `--miner.gasprice "500000000000"` (500 Gwei)
1. Submit some transactions to the network
1. `block_queue_blknum_submitting` should increase and `block_submission_attempt` is tallied
1. Childchain block submissions stay pending, not getting included in root chain blocks
1. `block_submit_stalled` alarm should raise
1. Set geth's `--miner.gasprice` to a very low value, e.g. `--miner.gasprice "100000000"` (0.1 Gwei)
1. Childchain block submissions start getting included in root chain blocks
1. `block_queue_blknum_submitted` should increase and `block_submission_success` is tallied
1. `block_submit_stalled` alarm should be cleared
